galley/pkg/ca/ca.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package ca
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/k8s/controller"
+	istioca "istio.io/istio/security/pkg/pki/ca"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+var (
+	// certChainID is the ID/name for the certificate chain file.
+	certChainID = "cert-chain.pem"
+	// privateKeyID is the ID/name for the private key file.
+	privateKeyID = "key.pem"
+	// rootCertID is the ID/name for the CA root certificate file.
+	rootCertID = "root-cert.pem"
+
+	galleyServiceAccountName = "istio-galley-service-account"
+
+	galleySecret = "istio." + galleyServiceAccountName
+)
+
+//GenerateSelfSignedCA generated self signed ca if istio.istio-galley-service-account does not exist
+func GenerateSelfSignedCA(caArgs *Args, workloadCertArgs *WorkloadCertArgs, kubeConfig, namespace, svcName string) error {
+
+	cs := createClientset(kubeConfig)
+	if !secretIsAlreadyExists(cs, namespace) {
+		err := createGalleySecret(cs, caArgs, workloadCertArgs, namespace, svcName)
+		if err != nil {
+			return fmt.Errorf("generate secret failed: %v", err)
+		}
+	} else {
+		return generateCAFilesAndWatchSecret(cs, caArgs, workloadCertArgs, namespace, svcName)
+	}
+	return nil
+}
```

Comment:
Is there any material different from sharing the self-signed CA and sharing the generated cert/key between webhook replicas? In either case, the current code can race between checking if the secret exists and creating it. To avoid this, we need to create the secret first and, if it already exists, use that instead.